Added first version of the Slack action

rwijngaa opened the pull request.

Commit:
Added first version of the Slack action

9037568 commented on bundles/action/org.openhab.action.slack/README.md:
```diff
@@ -0,0 +1,28 @@
+TBD: make the action available in the Designer
+
+
+###### Contact
```
This section should be removed.

9037568 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/Slack.java:
```diff
@@ -0,0 +1,106 @@
+ * It uses the Simple Slack Api from https://github.com/Ullink/simple-slack-api
+ *
+ * @author Rino van Wijngaarden (rino.van.wijngaarden@gmail.com)
+ *
```
@since 1.10.0

9037568 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/Slack.java:
```diff
@@ -0,0 +1,106 @@
+                    session.disconnect();
+                }
+            } catch (IOException e) {
+                logger.error("Error disonnecting Slack session", e);
```
"disconnecting"

9037568 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/Slack.java:
```diff
@@ -0,0 +1,106 @@
+                    session.disconnect();
+                }
+            } catch (IOException e) {
+                logger.error("Error disonnecting Slack session", e);
```
"disconnecting"

9037568 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/SlackActionService.java:
```diff
@@ -0,0 +1,69 @@
+/**
+ * This class registers an OSGi service for the Slack action.
+ *
+ * @author Rino van Wijngaarden
```
Should have the same value for author tags throughout.

9037568 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/SlackActionService.java:
```diff
@@ -0,0 +1,69 @@
+ * This class registers an OSGi service for the Slack action.
+ *
+ * @author Rino van Wijngaarden
+ *
```
@since 1.10.0

9037568 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/SlackActionService.java:
```diff
@@ -0,0 +1,69 @@
+            // check mandatory settings
+            if (StringUtils.isBlank(Slack.authToken)) {
+                throw new ConfigurationException("slack",
+                        "Parameter authToken is mandatory and must be configured. Please check your slack.cfg!");
```
Should not reference specific files here, as they'll vary by platform.

9037568 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/SlackActivator.java:
```diff
@@ -0,0 +1,55 @@
+ * Extension of the default OSGi bundle activator
+ *
+ * @author Rino van Wijngaarden
+ *
```
@since 1.10.0

9037568 reviewed (changes requested):
Just a few minor tweaks needed.

Please also update the openhab_default.cfg and create a wiki page for the action.

watou commented on bundles/action/org.openhab.action.slack/README.md:
```diff
@@ -0,0 +1,28 @@
+# OpenHab Slack Action
```
In the case of just the top-level header, please remove openHAB.  For overall structure of the README.md, please follow [this example](https://raw.githubusercontent.com/openhab/openhab1-addons/master/bundles/action/org.openhab.action.telegram/README.md) for heading titles and levels (the README.md will end up at docs.openhab.org eventually).

hakan42 commented on .gitignore:
```diff
@@ -18,6 +18,11 @@ distribution/openhabhome/etc/rrd4j/*.rrd
 features/*/src/main/history
 itemDump.txt
 
+#Eclipse files
```
Please remove changes to the gitignore file. If any are (globally) necessary, they should be in a seperate PR and go into the top-level gitignore file.

hakan42 commented on bundles/action/org.openhab.action.slack/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,40 @@
+Bundle-ClassPath: .,
+ lib/commons-logging-1.2.jar,
+ lib/gson-2.6.2.jar,
+ lib/httpclient-4.4.jar,
```
Do you really need all those libraries in those specific versions? If not, you could try to work with the ones provided from the ESH core.

I am only using openHAB 2.x, so I don't know what list of libraries are provide by 1.x though 😊

watou replied:
I am doubtful of the need to include commons-logging, but any code in this repo should work on the 1.x runtime (unless there is some extraordinary circumstance) so only the intersection of target platforms between 1 and 2 can be assumed.

hakan42 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/Slack.java:
```diff
@@ -0,0 +1,106 @@
+ */
+public class Slack {
+
+    private static final Logger logger = LoggerFactory.getLogger(Slack.class);
```
Logger should not be static.

watou replied:
The new convention in ESH/OS2 is to not have static loggers, but all the other 1.x add-ons have static loggers, so I would say this is OK.

hakan42 commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/SlackActivator.java:
```diff
@@ -0,0 +1,55 @@
+ */
+public final class SlackActivator implements BundleActivator {
+
+    private static Logger logger = LoggerFactory.getLogger(SlackActivator.class);
```
Logger should not be static.

hakan42 commented on bundles/action/org.openhab.action.slack/src/main/resources/readme.txt:
```diff
@@ -0,0 +1 @@
+Bundle resources go in here!
```
Please remove placeholder file if you have no required resources

watou replied:
Again, having this tiny file is common throughout the 1.x codebase.

hakan42 reviewed (changes requested):
A few little changes here and there, but overall a nice action.

Commit:
Review comments 9037568

rwijngaa commented:
Thanks for the review. I've pushed a new version!

watou commented on bundles/action/org.openhab.action.slack/README.md:
```diff
@@ -0,0 +1,49 @@
+# Slack Action
+The Slack action sends messages to a Slack ([https://slack.com/]) channel or Slack user from a rule or script.
```
Since this page will eventually be shown on docs.openhab.org, it has to conform to the markdown rendered used there as well as here on GitHub.  So every heading has to have a blank line above and below, and every code block surrounded by \`\`\` has to have a blank line before the beginning and a blank line after the end.

watou commented on bundles/action/org.openhab.action.slack/README.md:
```diff
@@ -0,0 +1,49 @@
+The Slack action sends messages to a Slack ([https://slack.com/]) channel or Slack user from a rule or script.
+
+## Prerequisites
+As described on the Slack page ([https://my.slack.com/services/new/bot]) you first need to create a bot.
```
Better would be `[Slack page](https://my.slack.com/services/new/bot)`.

watou commented on bundles/action/org.openhab.action.slack/src/main/java/org/openhab/action/slack/internal/SlackActivator.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public void start(BundleContext bc) throws Exception {
+        context = bc;
+        logger.info("Slack action has been started.");
```
Could this and the next logger statement be reduced to debug?

watou reviewed (changes requested):
A few more comments.  I really think you don't need any of the logging JARs and should only need to import the package for the logging API.  Please try to slim down the libraries as far as possible so the action still works in openHAB 1 runtime, and it ought to work under openHAB 2 as well (but please test this theory).

Commit:
Review comments: cleaned up dependencies and fixed Readme.md. Works in
OH2, but in OH1 i still have "Unsatisfied" (this was also the case
before the cleanup of dependencies)

watou commented:
> Works in  OH2, but in OH1 i still have "Unsatisfied" (this was also the case before the cleanup of dependencies)

Do you have a log and/or stacktrace from OH1 that could help diagnose which dependencies are not being resolved?  I think it would be very nice for this action to work on both OH1 and OH2 runtimes.

rwijngaa commented:
> Do you have a log and/or stacktrace from OH1 that could help diagnose which dependencies are not being resolved? I think it would be very nice for this action to work on both OH1 and OH2 runtimes
Well, i only see "The name 'sendToSlackUser(<XStringLiteralImpl>,<XStringLiteralImpl>)' cannot be resolved to an item or type." because of the "Unsatisfied" i guess.
Running "diag <bundleId>" tells me that all is good ("No unresolved constraints").  So, i need to look into this any further. Any hint, tips, tricks?

watou commented:
> Well, i only see "The name 'sendToSlackUser(,)' cannot be resolved to an item or type." because of the "Unsatisfied" i guess.

So do you mean to say that putting the action JAR in openHAB 1.8.3's addons folder, and then attempting to call `sendToSlackUser` from an openHAB script or rule with two otherwise valid arguments, produces some entry in openhab.log that contains the word "Unsatisfied"?  A snippet from any logs, inserted in a comment here like

\`\`\`
your log snippet
\`\`\`

would still be helpful.

Quite different would be complaints in openHAB Designer, which can be ignored for now.

rwijngaa commented:
ok, i started over with openhab1.8.3 and added this to demo.rules
```
                logInfo("Test slack", "before sendToSlack");
                sendToSlackUser("user test from openhab1.x", "openhab")
                logInfo("Test slack", "after sendToSlack");
```

And in the logging i only see this:
```
openhab.log:16:37:37.260 [INFO ] [penhab.model.script.Test slack:53   ] - before sendToSlack
```

So no error message and no "after sendToSlack" now .

On the osgi console i do a list and see:
```
74	Active		org.openhab.ui.webapp.mapviewrenderer			org.openhab.ui.webapp(bid=211)
75	Unsatisfied		org.openhab.action.slack.action			org.openhab.action.slack(bid=221)
76	Active		org.openhab.persistence.logging			org.openhab.persistence.logging(bid=222)
77	Active		org.openhab.binding.ntp.binding			org.openhab.binding.ntp(bid=223)
```

And diagnosing the bunde gives:
```
osgi> diag 221
file:/Users/rinovanwijngaarden/Projects/openhab1/openhab-1.8.3/addons/org.openhab.action.slack-1.10.0-SNAPSHOT.jar [221]
  No unresolved constraints.
```


and bundle info
```
osgi>
osgi> bundle 221
org.openhab.action.slack_1.10.0.201702231544 [221]
  Id=221, Status=<<LAZY>>    Data Root=/Users/rinovanwijngaarden/Projects/openhab1/openhab-1.8.3/server/configuration/org.eclipse.osgi/bundles/221/data
  "No registered services."
  No services in use.
  No exported packages
  Imported packages
    org.apache.commons.lang; version="2.6.0"<org.apache.commons.lang_2.6.0.v201205030909 [30]>
    org.apache.commons.logging; version="1.1.1"<org.slf4j.jcl_1.7.2.v20130115-1340 [215]>
    org.openhab.core.items; version="0.0.0"<org.openhab.core_1.8.3 [187]>
    org.openhab.core.library.items; version="0.0.0"<org.openhab.core.library_1.8.3 [189]>
    org.openhab.core.library.types; version="0.0.0"<org.openhab.core.library_1.8.3 [189]>
    org.openhab.core.scriptengine.action; version="0.0.0"<org.openhab.core.scriptengine_1.8.3 [192]>
    org.openhab.core.types; version="0.0.0"<org.openhab.core_1.8.3 [187]>
    org.osgi.framework; version="1.7.0"<org.eclipse.osgi_3.8.2.v20130124-134944 [0]>
    org.osgi.service.cm; version="1.4.0"<org.eclipse.osgi.services_3.3.100.v20120522-1822 [157]>
    org.osgi.service.component; version="1.2.0"<org.eclipse.osgi.services_3.3.100.v20120522-1822 [157]>
    org.slf4j; version="1.7.2"<org.slf4j.api_1.7.2.v20121108-1250 [214]>
  No fragment bundles
  Named class space
    org.openhab.action.slack; bundle-version="1.10.0.201702231544"[provided]
  No required bundles
```

So, the bundle is "Unsatisfied" but i can not see what cause this.

hakan42 commented:
@rwijngaa , do you still intend to work on this binding? Can I do something to help progress?
